azurerm/internal/services/storage/resource_arm_storage_account_customer_managed_key.go

```diff
@@ -0,0 +1,249 @@
+	storageClient := meta.(*clients.Client).Storage.AccountsClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	keyVaultId := d.Get("key_vault_id").(string)
+	keyName := d.Get("key_name").(string)
+	keyVersion := d.Get("key_version").(string)
+	storageAccountId := d.Get("storage_account_id").(string)
+	encryptionServices := true
+
+	storageAccountName, storageAccountResourceGroupName, err := azure.ParseNameAndResourceGroupFromResourceId(storageAccountId, "storageAccounts")
+	if err != nil {
+		return err
+	}
+
+	keyVaultAccountName, keyVaultResourceGroupName, err := azure.ParseNameAndResourceGroupFromResourceId(keyVaultId, "vaults")
+	if err != nil {
+		return err
+	}
+
+	// First check to see if the key vault is configured correctly or not
+	if !azure.KeyVaultIsSoftDeleteAndPurgeProtected(ctx, vaultClient, keyVaultId) {
+		return fmt.Errorf("Key Vault %q (Resource Group %q) is not configured correctly, please make sure that both 'soft_delete_enabled' and 'purge_protection_enabled' arguments are set to 'true'", keyVaultAccountName, keyVaultResourceGroupName)
+	}
+
+	pKeyVaultBaseURL, err := azure.GetKeyVaultBaseUrlFromID(ctx, vaultClient, keyVaultId)
+	if err != nil {
+		return fmt.Errorf("Error looking up Key Vault URI from Key Vault %q (Resource Group %q): %+v", keyVaultAccountName, keyVaultResourceGroupName, err)
+	}
+
+	props := storage.AccountUpdateParameters{
+		AccountPropertiesUpdateParameters: &storage.AccountPropertiesUpdateParameters{
+			Encryption: &storage.Encryption{
+				Services: &storage.EncryptionServices{
+					Blob: &storage.EncryptionService{
+						Enabled: utils.Bool(encryptionServices),
+					},
+					File: &storage.EncryptionService{
+						Enabled: utils.Bool(encryptionServices),
+					},
+				},
+				KeySource: storage.MicrosoftKeyvault,
+				KeyVaultProperties: &storage.KeyVaultProperties{
+					KeyName:     utils.String(keyName),
+					KeyVersion:  utils.String(keyVersion),
+					KeyVaultURI: utils.String(pKeyVaultBaseURL),
+				},
+			},
+		},
+	}
+
+	_, err = storageClient.Update(ctx, storageAccountResourceGroupName.(string), storageAccountName.(string), props)
+	if err != nil {
+		return fmt.Errorf("Error updating Azure Storage Account %q (Resource Group %q) Customer Managed Key: %+v", storageAccountName, storageAccountResourceGroupName, err)
+	}
+
+	resourceId := fmt.Sprintf("%s/customerManagedKey", storageAccountId)
+	d.SetId(resourceId)
+
+	return resourceArmStorageAccountCustomerManagedKeyRead(d, meta)
+}
+
+func resourceArmStorageAccountCustomerManagedKeyRead(d *schema.ResourceData, meta interface{}) error {
+	storageClient := meta.(*clients.Client).Storage.AccountsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	storageAccountId := d.Get("storage_account_id").(string)
+	keyVaultId := d.Get("key_vault_id").(string)
+
+	d.Set("storage_account_id", storageAccountId)
+	d.Set("key_vault_id", keyVaultId)
```

Comment:
we can't assume these exist during a read - we'll need to parse them from the ID/look them up